lgc/patch/PatchWorkarounds.cpp

```diff
@@ -0,0 +1,141 @@
+#include "llvm/IR/IntrinsicsAMDGPU.h"
+#include "llvm/Support/Debug.h"
+
+#define DEBUG_TYPE "lgc-patch-workarounds"
+
+using namespace lgc;
+using namespace llvm;
+
+namespace lgc {
+
+// =====================================================================================================================
+// Define static members (no initializer needed as LLVM only cares about the address of ID, never its value).
+char PatchWorkarounds::ID;
+
+// =====================================================================================================================
+// Pass creator, creates the pass of LLVM patching operations for peephole optimizations.
+FunctionPass *createPatchWorkarounds() {
+  return new PatchWorkarounds();
+}
+
+PatchWorkarounds::PatchWorkarounds() : FunctionPass(ID) {
+}
+
+// =====================================================================================================================
+// Executes this LLVM pass on the specified LLVM function.
+//
+// @param [in/out] function : Function that we will add workarounds.
+bool PatchWorkarounds::runOnFunction(Function &function) {
+  LLVM_DEBUG(dbgs() << "Run the pass Patch-Workarounds\n");
+
+  m_pipelineState = getAnalysis<PipelineStateWrapper>().getPipelineState(function.getParent());
+  m_context = &function.getContext();
+  m_builder = std::make_unique<IRBuilder<>>(*m_context);
+  m_processed.clear();
+
+  m_changed = false;
+
+  // At present we only apply a single workaround in this pass. Makes sense to check here an early exit the pass if not
+  // required
+  if (!m_pipelineState->getOptions().disableImageResourceCheck &&
+      m_pipelineState->getTargetInfo().getGpuWorkarounds().gfx10.waFixBadImageDescriptor)
+    visit(function);
+
+  return m_changed;
+}
+
+// =====================================================================================================================
+// Visits "call" instruction.
+//
+// @param callInst : The instruction
+void PatchWorkarounds::visitCallInst(CallInst &callInst) {
+  Function *const calledFunc = callInst.getCalledFunction();
+  if (!calledFunc)
+    return;
+
+  auto name = calledFunc->getName();
+
+  const bool isImage = name.startswith("llvm.amdgcn.image.");
+
+  // A buffer descriptor may be incorrectly given when it should be an image descriptor, we need to fix it to valid
+  // buffer type (0) to make hardware happily ignore it. This is to check and fix against buggy applications which
+  // declares a image descriptor in shader but provide a buffer descriptor in driver. Note this only applies to gfx10.1.
+  if (isImage) {
+    // Look for 8 dword resource descriptor argument
+    for (Value *arg : callInst.args()) {
+      if (auto vecTy = dyn_cast<FixedVectorType>(arg->getType())) {
+        if (vecTy->getNumElements() == 8 && vecTy->getElementType()->isIntegerTy(32)) {
+          if (isa<UndefValue>(arg))
+            // We don't need to worry if the value is actually undef
+            break;
```

Comment:
Did you intend to break here instead of continuing? Could you add a loop-level comment describing which image instructions this is looking for?

Reply:
Yes, break was intentional. The (not exactly ideal) scenario is that we look for an 8-dword resource descriptor argument. All paths that find this 8-dword desc lead to a break out of the loop (since there will only be one).
If the resource descriptor is undef, there's no point applying the workaround to the descriptor / or the instruction. This scenario only occurs in reality for tests (common practice to use an undef argument to make writing the test easier).

Reply:
I think you can incorporate your comment into the patch itself something like this is just for the case that is common in manually written test-case blabla... so it is more easy to understand. because it will be weird if we really meet with undef resource descriptor.

Reply:
Agreed. Done.